tests/koji/test_koji_vmi.py

```diff
@@ -0,0 +1,204 @@
+        list(source)
+
+    assert "Virtual machine image build not found in koji: notexist-1.2.3" in str(
+        exc_info.value
+    )
+
+
+def test_koji_not_vmi(fake_koji, koji_dir):
+    """Koji source referencing an invalid build will fail."""
+
+    source = Source.get(
+        "koji:https://koji.example.com/?vmi_build=foobuild-1.0-1", basedir=koji_dir
+    )
+
+    fake_koji.build_data[1234] = {
+        "id": 1234,
+        "name": "foobuild",
+        "version": "1.0",
+        "release": "1",
+        "nvr": "foobuild-1.0-1",
+        "completion_time": "2022-12-20 12:30:25",
+    }
+    fake_koji.build_data["foobuild-1.0-1"] = fake_koji.build_data[1234]
+
+    with raises(ValueError) as exc_info:
+        list(source)
+
+    assert (
+        "Build foobuild-1.0-1 not recognized as a virtual machine image build"
+        in str(exc_info.value)
+    )
+
+
+def test_koji_vmis(fake_koji, koji_dir):
+    """Koji source yields requested Virtual Machine Images."""
+
+    nvr = "foobuild-1.0-1"
+    r, v, n = nvr[-1::-1].split("-", 2)
+    name = n[-1::-1]
+    version = v[-1::-1]
+    release = r[-1::-1]
+
+    source = Source.get(
+        "koji:https://koji.example.com/?vmi_build=%s" % nvr, basedir=koji_dir
+    )
+
+    # It should not have done anything yet (lazy loading)
+    assert not fake_koji.last_url
+
+    # Insert some data
+    fake_koji.build_data[1234] = {
+        "id": 1234,
+        "name": name,
+        "version": version,
+        "release": release,
+        "nvr": nvr,
+        "completion_time": "2022-12-20 12:30:25",
+        "extra": {"typeinfo": {"image": {}}},
+    }
+    fake_koji.build_data[nvr] = fake_koji.build_data[1234]
+
+    imgs = {
+        "foo-1.0-1.img": "generic",
+        "foo-1.0-1.raw.xz": "raw-xz",
+        "foo-1.0-1.vhd.xz": "vhd-compressed",
+    }
+
+    archives = []
+    for fname, tname in imgs.items():
+        archives.append(
+            {
+                "btype": "image",
+                "filename": fname,
+                "checksum": "00000000000000000000000000000000",
+                "extra": {"image": {"arch": "x86_64"}},
+                "type_name": tname,
+                "build_id": 1234,
+                "nvr": nvr,
+            }
+        )
+
+    fake_koji.insert_archives(archives=archives, build_nvr=nvr)
+
+    # raise Exception(fake_koji.build_data)
+
+    # Eagerly fetch
+    items = list(source)
+
+    # It should have returned push items for the modules
+    assert len(items) == 3
+
+    items = sorted(items, key=lambda pi: pi.name)
+
+    klasses = [VMIPushItem, AmiPushItem, VHDPushItem]
+    index = 0
+
+    for klass in klasses:
+        fname = sorted(list(imgs.keys()))[index]
+
+        rel_klass = AmiRelease if klass == AmiPushItem else VMIRelease
+
+        rel_obj = rel_klass(
+            arch="x86_64",
+            date="2022-12-20",
+            product=name.upper(),
+            version=version,
+            respin=int(r[-1::-1]),
```

Comment:
Again, I think it is better to just hardcode the value for comparison so it can be seen at a glance whether or not the value makes sense.

Reply:
Sure!

Reply:
Fixed! It makes more sense to hardcode as you suggested, thanks a lot!